File: problems/0295-find-median-from-data-stream/go/solution.go
package solution

import "container/heap"

type minH []int

func (h minH) Len() int           { return len(h) }
func (h minH) Less(i, j int) bool { return h[i] < h[j] }
func (h minH) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *minH) Push(x any)        { *h = append(*h, x.(int)) }
func (h *minH) Pop() any {
	o := *h
	n := len(o)
	x := o[n-1]
	*h = o[:n-1]
	return x
}

type maxH []int

func (h maxH) Len() int           { return len(h) }
func (h maxH) Less(i, j int) bool { return h[i] > h[j] }
func (h maxH) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *maxH) Push(x any)        { *h = append(*h, x.(int)) }
func (h *maxH) Pop() any {
	o := *h
	n := len(o)
	x := o[n-1]
	*h = o[:n-1]
	return x
}

// Find Median from Data Stream — https://leetcode.com/problems/find-median-from-data-stream/
type MedianFinder struct {
	small *maxH // lower half (max-heap)
	large *minH // upper half (min-heap)
}

func NewMedianFinder() *MedianFinder {
	return &MedianFinder{small: &maxH{}, large: &minH{}}
}

func (m *MedianFinder) AddNum(num int) {
	heap.Push(m.small, num)
	// move the largest of small into large to enforce ordering
	heap.Push(m.large, heap.Pop(m.small))
	// keep small at least as large as large
	if m.large.Len() > m.small.Len() {
		heap.Push(m.small, heap.Pop(m.large))
	}
}

func (m *MedianFinder) FindMedian() float64 {
	if m.small.Len() > m.large.Len() {
		return float64((*m.small)[0])
	}
	return float64((*m.small)[0]+(*m.large)[0]) / 2
}

func medianFinderOps(ops []string, args [][]int) []float64 {
	mf := NewMedianFinder()
	out := []float64{}
	for i, op := range ops {
		switch op {
		case "MedianFinder":
			mf = NewMedianFinder()
		case "addNum":
			mf.AddNum(args[i][0])
		case "findMedian":
			out = append(out, mf.FindMedian())
		}
	}
	return out
}

File: problems/0295-find-median-from-data-stream/typescript/solution.ts
// A binary heap ordered by a comparator (`less(a, b)` true => a is higher priority).
class Heap {
  private data: number[] = [];
  constructor(private less: (a: number, b: number) => boolean) {}

  size(): number {
    return this.data.length;
  }

  peek(): number {
    return this.data[0];
  }

  push(val: number): void {
    this.data.push(val);
    let i = this.data.length - 1;
    while (i > 0) {
      const p = (i - 1) >> 1;
      if (!this.less(this.data[i], this.data[p])) break;
      [this.data[p], this.data[i]] = [this.data[i], this.data[p]];
      i = p;
    }
  }

  pop(): number {
    const top = this.data[0];
    const last = this.data.pop();
    if (last !== undefined && this.data.length > 0) {
      this.data[0] = last;
      let i = 0;
      const n = this.data.length;
      while (true) {
        let best = i;
        const l = 2 * i + 1;
        const r = 2 * i + 2;
        if (l < n && this.less(this.data[l], this.data[best])) best = l;
        if (r < n && this.less(this.data[r], this.data[best])) best = r;
        if (best === i) break;
        [this.data[best], this.data[i]] = [this.data[i], this.data[best]];
        i = best;
      }
    }
    return top;
  }
}

// Find Median from Data Stream — https://leetcode.com/problems/find-median-from-data-stream/
class MedianFinder {
  private small = new Heap((a, b) => a > b); // max-heap: lower half
  private large = new Heap((a, b) => a < b); // min-heap: upper half

  addNum(num: number): void {
    this.small.push(num);
    this.large.push(this.small.pop());
    if (this.large.size() > this.small.size()) {
      this.small.push(this.large.pop());
    }
  }

  findMedian(): number {
    if (this.small.size() > this.large.size()) return this.small.peek();
    return (this.small.peek() + this.large.peek()) / 2;
  }
}

export function medianFinderOps(ops: string[], args: number[][]): number[] {
  let mf = new MedianFinder();
  const out: number[] = [];
  for (let i = 0; i < ops.length; i++) {
    if (ops[i] === "MedianFinder") mf = new MedianFinder();
    else if (ops[i] === "addNum") mf.addNum(args[i][0]);
    else if (ops[i] === "findMedian") out.push(mf.findMedian());
  }
  return out;
}
